File: messagebus/src/vespa/messagebus/routing/routeparser.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "errordirective.h"
#include "policydirective.h"
#include "routedirective.h"
#include "routeparser.h"
#include "tcpdirective.h"
#include "verbatimdirective.h"
#include <vespa/vespalib/util/stringfmt.h>

using vespalib::stringref;

namespace mbus {

bool
RouteParser::isWhitespace(char c)
{
    return c == ' ' || c == '\f' || c == '\n' || c == '\r' || c == '\t';
}

IHopDirective::SP
RouteParser::createRouteDirective(const stringref &str)
{
    return IHopDirective::SP(new RouteDirective(str));
}

IHopDirective::SP
RouteParser::createTcpDirective(const stringref &str)
{
    size_t posP = str.find(":");
    if (posP == string::npos || posP == 0) {
        return IHopDirective::SP(); // no host
    }
    size_t posS = str.find("/", posP);
    if (posS == string::npos || posS == posP + 1) {
        return IHopDirective::SP(); // no port
    }
    return IHopDirective::SP(new TcpDirective(str.substr(0, posP),
                                              atoi(str.substr(posP + 1, posS - 1).c_str()),
                                              str.substr(posS + 1)));
}

IHopDirective::SP
RouteParser::createPolicyDirective(const stringref &str)
{
    size_t pos = str.find(":");
    if (pos == string::npos) {
        return IHopDirective::SP(new PolicyDirective(str, ""));
    }
    return IHopDirective::SP(new PolicyDirective(str.substr(0, pos), str.substr(pos + 1)));
}

IHopDirective::SP
RouteParser::createVerbatimDirective(const stringref &str)
{
    return IHopDirective::SP(new VerbatimDirective(str));
}

IHopDirective::SP
RouteParser::createErrorDirective(const stringref &str)
{
    return IHopDirective::SP(new ErrorDirective(str));
}

IHopDirective::SP
RouteParser::createDirective(const stringref &str)
{
    if (str.size() > 2 && str[0] == '[') {
        return createPolicyDirective(str.substr(1, str.size() - 2));
    }
    return createVerbatimDirective(str);
}

Hop
RouteParser::createHop(stringref str)
{
    if (str.empty()) {
        return Hop().addDirective(createErrorDirective("Failed to parse empty string."));
    }
    size_t len = str.size();
    if (len > 1 && str[0] == '?') {
        Hop hop = createHop(str.substr(1));
        hop.setIgnoreResult(true);
        return hop;
    }
    if (len > 4 && str.substr(0, 4) == "tcp/") {
        IHopDirective::SP tcp = createTcpDirective(str.substr(4));
        if (tcp.get() != nullptr) {
            return Hop().addDirective(tcp);
        }
    }
    if (len > 6 && str.substr(0, 6) == "route:") {
        return Hop().addDirective(createRouteDirective(str.substr(6)));
    }
    Hop ret;
    for (size_t from = 0, at = 0, depth = 0; at <= len; ++at) {
        if (at == len || (depth == 0 && str[at] == '/')) {
            if (depth > 0) {
                return Hop().addDirective(createErrorDirective(
                                "Unexpected token '': syntax error"));
            }
            ret.addDirective(createDirective(str.substr(from, at - from)));
            from = at + 1;
        } else if (isWhitespace(str[at]) && depth == 0) {
            return Hop().addDirective(createErrorDirective(
                            vespalib::make_string(
                                    "Failed to completely parse '%s'.",
                                    str.c_str())));
        } else if (str[at] == '[') {
            ++depth;
        } else if (str[at] == ']') {
            if (depth == 0) {
                return Hop().addDirective(createErrorDirective(
                                "Unexpected token ']': syntax error"));
            }
            --depth;
        }
    }
    return ret;
}

Route
RouteParser::createRoute(stringref str)
{
    Route ret;
    for (size_t from = 0, at = 0, depth = 0; at <= str.size(); ++at) {
        if (at == str.size() || (depth == 0 && isWhitespace(str[at]))) {
            if (from < at - 1) {
                Hop hop = createHop(str.substr(from, at - from));
                if (hop.hasDirectives() &&
                    hop.getDirective(0)->getType() == IHopDirective::TYPE_ERROR)
                {
                    return std::move(Route().addHop(std::move(hop)));
                }
                ret.addHop(std::move(hop));
            }
            from = at + 1;
        } else if (str[at] == '[') {
            ++depth;
        } else if (str[at] == ']') {
            --depth;
        }
    }
    return ret;
}

} // mbus

File: messagebus/src/main/java/com/yahoo/messagebus/routing/RouteParser.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.messagebus.routing;

/**
 * This replaces the incredibly slow javacc RouteParser.jj. It is a has its c++ sibling and
 * the implementation is a a copy of the C++ version.
 * @author baldersheim
 * @since 5.2
 */

public class RouteParser {
    private final String routeText;
    RouteParser(String route) {
        this.routeText = route;
    }

    Route route() {
        Route route = new Route();
        for (int from = 0, at = 0, depth = 0; at <= routeText.length(); ++at) {
            if (at == routeText.length() || ((depth == 0) && Character.isWhitespace(routeText.charAt(at)))) {
                if (from < at) {
                    Hop hop = createHop(routeText.substring(from, at));
                    if (hop.hasDirectives() && hop.getDirective(0) instanceof ErrorDirective) {
                        return new Route().addHop(new Hop().addDirective(hop.getDirective(0)));
                    }
                    route.addHop(hop);
                }
                from = at + 1;
            } else if ((routeText.charAt(at) == '(') || (routeText.charAt(at) == '[') ) {
                ++depth;
            } else if ((routeText.charAt(at) == ')') || (routeText.charAt(at) == ']')) {
                --depth;
            }
        }
        return route;
    }
    private static Hop createHop(String s) {
        final int len = s.length();
        if (len == 0) {
            return new Hop().addDirective(createErrorDirective("Failed to parse empty string."));
        } else if (len > 1 && (s.charAt(0) == '?')) {
            return createHop(s.substring(1, len)).setIgnoreResult(true);
        } else if (len > 4 && s.charAt(0) == 't' && s.charAt(1) == 'c' && s.charAt(2) == 'p' && s.charAt(3) == '/') {
            HopDirective directive = createTcpDirective(s.substring(4, len));
            if (directive != null) {
                return new Hop().addDirective(directive);
            }
        } else if (len > 6 && s.charAt(0) == 'r' && s.charAt(1) == 'o' && s.charAt(2) == 'u'
                           && s.charAt(3) == 't' && s.charAt(4) == 'e' && s.charAt(5) == ':') {
            return new Hop().addDirective(createRouteDirective(s.substring(6, len)));
        }
        Hop hop = new Hop();
        for (int from = 0, at = 0, depth = 0; at <= len; ++at) {
            if (at == len) {
                if (depth > 0) {
                    return new Hop().addDirective(createErrorDirective("Unterminated '[' in '" + s + "'"));
                }
                hop.addDirective(createDirective(s.substring(from, at)));
                from = at + 1;
            } else {
                char c = s.charAt(at);
                if (Character.isWhitespace(c) && depth == 0) {
                    return new Hop().addDirective(createErrorDirective("Failed to completely parse '" + s + "'."));
                } else if ((depth == 0 && c == '/')) {
                    hop.addDirective(createDirective(s.substring(from, at)));
                    from = at + 1;
                } else if (c == '[') {
                    ++depth;
                } else if (c == ']') {
                    if (depth == 0) {
                        return new Hop().addDirective(createErrorDirective("Unexpected token ']' in '" + s + "'"));
                    }
                    --depth;
                }
            }
        }
        return hop;
    }
    private static HopDirective createErrorDirective(String s) {
        return new ErrorDirective(s);
    }
    private static HopDirective createTcpDirective(String s) {
        int posP = s.indexOf(':');
        if (posP <= 0) {
            return null;
        }
        int posS = s.indexOf('/', posP+1);
        if (posS <= posP + 1) {
            return null;
        }
        return new TcpDirective(s.substring(0, posP),
                Integer.valueOf(s.substring(posP + 1, posS)),
                s.substring(posS + 1));
    }
    private static HopDirective createRouteDirective(String s) {
        return new RouteDirective(s);
    }
    private static HopDirective createDirective(String s) {
        return (s.length() > 2 && s.charAt(0) == '[')
                ? ((s.charAt(s.length() - 1) == ']')
                    ? createPolicyDirective(s.substring(1, s.length() - 1))
                    : createErrorDirective("Unterminated '[' in '" + s + "'"))
                : createVerbatimDirective(s);

    }
    private static HopDirective createPolicyDirective(String s)
    {
        int pos = s.indexOf(':');
        return (pos == -1)
                ? new PolicyDirective(s, "")
                : new PolicyDirective(s.substring(0, pos), s.substring(pos + 1).trim());
    }

    private static HopDirective createVerbatimDirective(String s) {
        return new VerbatimDirective(s);
    }

}
